src/core/ai/llm/messages/utils.ts

```diff
@@ -149,3 +149,88 @@ export function getFileData(filePart: {
+
+export interface FilteringConfig {
+    provider: string;
+    model?: string;
+}
+
+/**
+ * Filters message content based on LLM capabilities.
+ * Removes unsupported file attachments while preserving supported content.
+ * Uses model-specific validation when available, falls back to provider-level.
+ * @param messages Array of internal messages to filter
+ * @param config The LLM configuration (provider and optional model)
+ * @returns Filtered messages with unsupported content removed
+ */
+export function filterMessagesByLLMCapabilities(
+    messages: InternalMessage[],
+    config: FilteringConfig
+): InternalMessage[] {
+    try {
+        return messages.map((message) => {
+            // Only filter user messages with array content (multimodal)
+            if (message.role !== 'user' || !Array.isArray(message.content)) {
+                return message;
+            }
+
+            const filteredContent = message.content.filter((part) => {
+                // Keep text and image parts
+                if (part.type === 'text' || part.type === 'image') {
+                    return true;
+                }
+
+                // Filter file parts based on LLM capabilities
+                if (part.type === 'file' && part.mimeType) {
+                    const validation = validateFileForLLM(config, part.mimeType);
+                    return validation.isSupported;
+                }
+
+                return true; // Keep unknown part types
+            });
+
+            // If all content was filtered out, add a placeholder text
+            if (filteredContent.length === 0) {
+                const modelContext = config.model ? ` by ${config.model}` : ' by current LLM';
+                filteredContent.push({
+                    type: 'text',
+                    text: `[File attachment removed - not supported${modelContext}]`,
+                });
+            }
+
+            return {
+                ...message,
+                content: filteredContent,
+            };
+        });
+    } catch (error) {
+        // If filtering fails, return original messages to avoid breaking the flow
+        console.warn('Failed to filter messages by LLM capabilities:', error);
+        return messages;
+    }
+}
+
+/**
+ * @deprecated Use filterMessagesByLLMCapabilities instead
+ * Filters message content based on LLM provider capabilities.
+ */
+export function filterMessagesByCapabilities(
+    messages: InternalMessage[],
+    provider: string
+): InternalMessage[] {
+    return filterMessagesByLLMCapabilities(messages, { provider });
+}
+
+/**
+ * @deprecated Use filterMessagesByLLMCapabilities instead
+ * Filters message content based on specific model capabilities.
+ */
+export function filterMessagesByModelCapabilities(
+    messages: InternalMessage[],
+    provider: string,
+    model: string
+): InternalMessage[] {
+    return filterMessagesByLLMCapabilities(messages, { provider, model });
+}
```

Comment:
delete

to avoid this kind of generation, add the prompt 'don't worry about backward compatibility' for PR refactors